src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -29,27 +32,70 @@ namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
 
 internal abstract class LanguageServerProjectLoader
 {
-    protected readonly AsyncBatchingWorkQueue<ProjectToLoad> ProjectsToLoadAndReload;
+    private readonly AsyncBatchingWorkQueue<ProjectToLoad> _projectsToLoadAndReload;
 
     protected readonly ProjectSystemProjectFactory ProjectFactory;
     private readonly ProjectTargetFrameworkManager _targetFrameworkManager;
     private readonly ProjectSystemHostInfo _projectSystemHostInfo;
     private readonly IFileChangeWatcher _fileChangeWatcher;
-    private readonly IGlobalOptionService _globalOptionService;
+    protected readonly IGlobalOptionService GlobalOptionService;
     protected readonly ILoggerFactory LoggerFactory;
     private readonly ILogger _logger;
     private readonly ProjectLoadTelemetryReporter _projectLoadTelemetryReporter;
     private readonly BinlogNamer _binlogNamer;
-    private readonly ProjectFileExtensionRegistry _projectFileExtensionRegistry;
     protected readonly ImmutableDictionary<string, string> AdditionalProperties;
 
     /// <summary>
-    /// The list of loaded projects in the workspace, keyed by project file path. The outer dictionary is a concurrent dictionary since we may be loading
-    /// multiple projects at once; the key is a single List we just have a single thread processing any given project file. This is only to be used
-    /// in <see cref="LoadOrReloadProjectsAsync" /> and downstream calls; any other updating of this (like unloading projects) should be achieved by adding
-    /// things to the <see cref="ProjectsToLoadAndReload" />.
+    /// Guards access to <see cref="_loadedProjects"/>.
+    /// To keep the LSP queue responsive, <see cref="_gate"/> must not be held while performing design-time builds.
     /// </summary>
-    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(1);
+
+    /// <summary>
+    /// Guarded by <see cref="_gate"/>.
+    /// Maps the file path of a tracked project to the load state for the project.
+    /// Absence of an entry indicates the project is not tracked, e.g. it was never loaded, or it was unloaded.
+    /// <see cref="_gate"/> must be held when modifying the dictionary or objects contained in it.
+    /// </summary>
+    private readonly Dictionary<string, ProjectLoadState> _loadedProjects = [];
```

Comment:
doc what the keys are here.  it's non obvious.

Reply:
The doc says:
> Maps the file path of a tracked project to the load state for the project

Was there a different phrasing of this that would make it more clear?

Reply:
my brain must have missed that.  sorry.  in general, i wouldn't mind is being precise on path.  like is this some URI thing.  some path on disk?  some "special fake/virtual project path".  We have a lot of paths flying around, and if they are typed as string, being clear which one we mean would be helpful.

alternatively, use strong types for the varying path concept.s

Reply:
I could imagine introducing stuff like `record struct OnDiskFilePath(string)`, `record struct VirtualFilePath(string)` etc.

Reply:
Per comments from Jason I think we just want to doc carefully when the "virtual path' is used, and not introduce any value-object types for paths at this time.